numpy/lib/arraypad.py

```diff
@@ -1193,192 +786,128 @@ def pad(array, pad_width, mode, **kwargs):
-            newmat = _append_min(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'reflect':
-        for axis, (pad_before, pad_after) in enumerate(pad_width):
-            if narray.shape[axis] == 0:
-                # Axes with non-zero padding cannot be empty.
-                if pad_before > 0 or pad_after > 0:
-                    raise ValueError("There aren't any elements to reflect"
-                                     " in axis {} of `array`".format(axis))
-                # Skip zero padding on empty axes.
-                continue
-
-            # Recursive padding along any axis where `pad_amt` is too large
-            # for indexing tricks. We can only safely pad the original axis
-            # length, to keep the period of the reflections consistent.
-            if ((pad_before > 0) or
-                    (pad_after > 0)) and newmat.shape[axis] == 1:
+    }
+    unsupported_kwargs = set(kwargs) - set(allowed_kwargs[mode])
+    if unsupported_kwargs:
+        raise ValueError("unsupported keyword arguments for mode '{}': {}"
+                         .format(mode, unsupported_kwargs))
+
+    stat_functions = {"maximum": np.max, "minimum": np.min,
+                      "mean": np.mean, "median": np.median}
+
+    # Create array with final shape and original values
+    # (padded area is undefined)
+    padded, _ = _pad_simple(array, pad_width)
+    # And prepare iteration over all dimensions
+    # (zipping may be more readable than using enumerate)
+    axes = range(padded.ndim)
+
+    if array.size == 0 and mode not in {"constant", "empty"}:
+        # Deal with special case: only modes "constant" and "empty" can extend
+        # empty axes, all other modes depend on `array` not being empty.
+        for axis, index_pair in zip(axes, pad_width):
+            if (
+                array.shape[axis] == 0
+                and (index_pair[0] > 0 or index_pair[1] > 0)
+            ):
+                raise ValueError(
+                    "can't extend empty axis {} using modes other than "
+                    "'constant' or 'empty'".format(axis)
+                )
+        # If pad values in empty axis were zero as well, then _pad_simple
+        # already returned the correct result
+        return padded
+
+    if mode == "constant":
+        values = kwargs.get("constant_values", 0)
+        values = _as_pairs(values, padded.ndim, assert_number=True)
+        for axis, index_pair, value_pair in zip(axes, pad_width, values):
+            _set_pad_area(padded, axis, index_pair, value_pair)
+
+    elif mode == "edge":
+        for axis, index_pair in zip(axes, pad_width):
+            edge_pair = _get_edges(padded, axis, index_pair)
+            _set_pad_area(padded, axis, index_pair, edge_pair)
+
+    elif mode == "linear_ramp":
+        end_values = kwargs.get("end_values", 0)
+        end_values = _as_pairs(end_values, padded.ndim, assert_number=True)
```

Comment:
`assert_number` changes the beaviour of `np.pad([Fraction(10)], (2, 2), mode='linear_ramp', end_values=Fraction(1,2))`, which previously mostly did the right thing, to erroring.

Reply:
Good catch! Removing `assert_number=True` restores the old behavior,